File: app/src/main/java/com/github/batulovandrey/unofficialurbandictionary/model/PopularWordsModel.java
package com.github.batulovandrey.unofficialurbandictionary.model;

import com.github.batulovandrey.unofficialurbandictionary.UrbanDictionaryApp;
import com.github.batulovandrey.unofficialurbandictionary.data.PopularWords;
import com.github.batulovandrey.unofficialurbandictionary.presenter.PopularWordsPresenter;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

/**
 * @author Andrey Batulov on 31/10/2017
 */

public class PopularWordsModel {

    @Inject
    PopularWords mWords;

    private final PopularWordsPresenter mPopularWordsPresenter;
    private Map<String, List<String>> mDictionary;

    public PopularWordsModel(PopularWordsPresenter popularWordsPresenter) {
        UrbanDictionaryApp.getNetComponent().inject(this);
        mPopularWordsPresenter = popularWordsPresenter;
        mDictionary = mWords.getDictionary();
    }

    public ArrayList<String> getAlphabet() {
        ArrayList<String> alphabet = new ArrayList<>();
        for (Map.Entry<String, List<String>> map : mDictionary.entrySet()) {
            alphabet.add(map.getKey());
        }
        return alphabet;
    }

    public ArrayList<String> getWords(String letter) {
        ArrayList<String> words = new ArrayList<>();
        for (Map.Entry<String, List<String>> map : mDictionary.entrySet()) {
            if (map.getKey().equals(letter)) {
                words.addAll(map.getValue());
            }
        }
        return words;
    }
}

File: app/src/main/java/com/github/batulovandrey/unofficialurbandictionary/model/PopularWordsModel.kt
package com.github.batulovandrey.unofficialurbandictionary.model

import com.github.batulovandrey.unofficialurbandictionary.UrbanDictionaryApp
import com.github.batulovandrey.unofficialurbandictionary.data.PopularWords
import com.github.batulovandrey.unofficialurbandictionary.presenter.PopularWordsPresenter
import java.util.*
import javax.inject.Inject

/**
 * @author Andrey Batulov on 22/12/2017
 */

class PopularWordsModel(private val mPopularWordsPresenter: PopularWordsPresenter) {

    @Inject
    lateinit var mWords: PopularWords

    private val mDictionary: Map<String, List<String>>

    val alphabet: ArrayList<String>
        get() {
            val alphabet = ArrayList<String>()
            for ((key) in mDictionary) {
                alphabet.add(key)
            }
            return alphabet
        }

    init {
        UrbanDictionaryApp.getNetComponent().inject(this)
        mDictionary = mWords.dictionary
    }

    fun getWords(letter: String): ArrayList<String> {
        val words = ArrayList<String>()
        for ((key, value) in mDictionary) {
            if (key == letter) {
                words.addAll(value)
            }
        }
        return words
    }
}